### DISCUSSION TITLE
emacs + netrepl setup

### DISCUSSION THREAD

**Discussion by williewillus** [janet-lang/janet#1318] [General]
Wondering if anyone's gotten a good setup going for emacs and netrepl?

All the repl setups I see around the internet have you starting the Janet process from emacs, but I'm making a game. Displaying a window/drawing graphics doesn't really fit well to being started from emacs, so I'd rather start the game independently, it spins up a netrepl on a fiber, then I connect to it from emacs.

**Comment by sogaiu** [comment]
There are a couple of things I've tried building:

* https://github.com/sogaiu/a-janet-spork-client
* https://github.com/sogaiu/snr

The terribly-named former one (a-janet-spork-client) might possibly be enough (or a base to work from may be?).  It doesn't support some bits like [`source` and `cancel`](https://github.com/janet-lang/janet/blob/46b34833c24bebd262d833dcb5b855fc45578923/src/boot/boot.janet#L2620-L2626) though [1].

The latter one (snr) starts a proxy so that the communication to / from an emacs process can use standard input and output instead of sockets.  snr is newer and I haven't worked with it as much, but FWIW.

---

[1] Olical's conjure has some level of support for `source` AFAICT though, so if interested in that sort of functionality, [this](https://github.com/Olical/conjure/blob/58c46d1f4999679659a5918284b574c266a7ac83/fnl/conjure/client/janet/netrepl.fnl#L58-L59) might be worth looking at.

**Comment by williewillus** [comment]
snr doesn't seem to work with local jpm trees, it expects spork installed in a global location (which I don't want to do). Is there a way to force it to use the jpm tree of the current project?

**Comment by williewillus** [comment]
I installed spork globally, and it still doesn't work -- now it just hangs forever.
The same happens if I try to use `netrepl/client` manually on the terminal to connect. I must be missing something about how to set up the server properly?

It's a bit disappointing as a newcomer how nonexistent emacs netrepl support is :(

**Comment by sogaiu** [comment]
As mentioned in Zulip, it looks like better behavior can be obtained if the game loop yields (e.g. by using a call to `ev/sleep`).

IIUC, the netrepl server runs in a fiber and the game loop lives in the same thread on a separate fiber and it seems to work better if the two cooperate by taking turns.

I had success connecting and evaluating using a-janet-spork-client as well as [conjure](https://github.com/Olical/conjure/) once an appropriate `ev/sleep` had been inserted into the game loop.

---

> snr doesn't seem to work with local jpm trees, it expects spork installed in a global location (which I don't want to do). Is there a way to force it to use the jpm tree of the current project?

There isn't currently.  I'll try to think about this though.  Thanks for the idea.

My intent in mentioning snr was more along the lines of pointing out some code that might be adapted.  Sorry if that didn't come across well (^^;

**Reply by dawranliou** [reply]
> I had success connecting and evaluating using a-janet-spork-client as well as [conjure](https://github.com/Olical/conjure/) once an appropriate ev/sleep had been inserted into the game loop.

Could you expand on this more?  Please bear with me as I'm still getting familiar with cooperative concurrent programming.  I don't know how to have my game loop yield control back to the `*ajsc-repl*` so I can evaluate code while the game loop is going.  

Here's a part of my game code (using jaylib) for illustration:

```janet
(defn start []
  (init-window 800 450 "game")
  (init-game)
  (set-target-fps 60)

  (while (not (window-should-close))
    (ev/sleep 0)  # <-- it doesn't seem to yield control back
    (update-game)
    (draw-game))

  (close-window))

(defn main [&]
  (netrepl/server-single))
```

I load and start the netrepl server by running `jpm -l janet game.janet`.  From my Emacs, I can connect to the netrepl server with `M-x ajsc`.  But then, when I load the buffer and execute `(start)`, I can no longer evaluate things in the `*ajsc-repl*` buffer.  Please let me know if you spot anything.

**Reply by sogaiu** [reply]
[This demo repository](https://github.com/sogaiu/jaylib-netrepl-demo) roughly recreates what I did (it's based on code that was made available via Zulip -- but that code does not appear to be available any more).

Regarding the posted code, I'm not sure that evaluating `(start)` in the manner described above is likely to work with the previously presented setup as it looks like `start` doesn't return until the `while` loop is exited.

**Reply by dawranliou** [reply]
Thanks for sharing the demo repository!  You are right about invoking the `(start)` function.  I actually need to invoke it with `(ev/call start)` so it can be scheduled on the event-loop.  Also, I learned from your repo that I need to provide the current fiber environment to the netrepl server.  Thanks again!  Have a great one!

**Comment by sogaiu** [comment]
> snr doesn't seem to work with local jpm trees, it expects spork installed in a global location (which I don't want to do). Is there a way to force it to use the jpm tree of the current project?

I thought about this a bit more and it is possible to do this currently for some setups.  Below is a description of something that worked here.

In a project directory where one has already done `jpm -l deps` (and one of the dependencies is spork), apply the `JANET_PATH` environment variable.

1. In one terminal (or via one's program), start a netrepl server, e.g.

    ```
    JANET_PATH=$(pwd)/jpm_tree/lib janet -e '(import spork/netrepl) (netrepl/server)'
    ```

2. In another terminal (or via snr in emacs [1]), start the proxy, e.g.
 
    ```
    JANET_PATH=$(pwd)/jpm_tree/lib janet netrepl-proxy.janet
    ```

This arrangement worked for me.

I tried evaluation via the terminal and also via emacs.

Via the terminal I saw prompts -- via emacs I didn't see prompts [2] (`M-x snr` resulted in a buffer with no prompts), but evaluation did work.

Note that, like a-janet-spork-client, snr doesn't support `source` or `cancel` either, so the former might be a better fit for the originally described setup.

---

[1] Note that `JANET_PATH` needs to be set appropriately for the emacs process as well.

Probably known, but for future readers, one way to do this is to start emacs like `JANET_PATH=... emacs` (with an appropriate value for `...`).

Another way is to evaluate the emacs-lisp form `(setenv "JANET_PATH" ...)` from within emacs before `M-x snr`.

Both of these worked for me.

[2] This is the way snr is currently.
